### PR TITLE
fix(duckdb): workaround for duckdb Map NULL bugs

### PR DESCRIPTION
**Pull request by NickCrews**
partially addresses https://github.com/ibis-project/ibis/issues/8632

### CODE REVIEW

**Comment by NickCrews** [comment]
Once that xfail test starts passing, then we could
- check for the duckdb version, and if we are running a broken version, then do the workaround, otherwise short-circuit to the simple path
- something else?

**Comment by NickCrews** [comment]
errr, this is probably the wrong way around this, I should be fixing this at the duckdb level, not the global ibis level. Otherwise we miss MapValues that are created via other APIs, like ibis.literal()

File: ibis/backends/tests/test_map.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
 from __future__ import annotations
 
+import duckdb
```
Comment (cpcloud) [inline ibis/backends/tests/test_map.py:3]: You cannot import this here. The tests in `ibis/backends/tests` must be runnable by arbitrary backends without having the dependencies for all backends installed.

For example, let's say I have the polars backend dependencies installed, but not the duckdb backend dependencies and I want to run the backend test suite using polars. Importing `duckdb` here (or the the dependencies for any non-polars backend) will immediately fail with an `ImportError`.

If you need to do something specific to a backend, put that code in `ibis/backend/${specific_backend}/tests`.

File: ibis/backends/tests/test_map.py [hunk 2]
```diff
@@ -26,6 +28,34 @@
...
+)
+def test_map_nulls(con, kv):
+    k, v = kv
+    k = ibis.literal(k, type="array<string>")
+    v = ibis.literal(v, type="array<string>")
+    m = ibis.map(k, v)
+    assert con.execute(m) is None
+
+
+@pytest.mark.xfail(
```
Comment (cpcloud) [inline ibis/backends/tests/test_map.py:49]: Move this test to `ibis/backends/tests/test_client.py` or if there's another file that makes sense, then there.

File: ibis/expr/types/structs.py [hunk 3]
```diff
@@ -75,6 +75,11 @@ def struct(
     """
     import ibis.expr.operations as ops
 
+    if isinstance(value, StructValue):
+        if type is not None:
+            return value.cast(type)
+        return value
```
Comment (cpcloud) [inline ibis/expr/types/structs.py:81]: This functionality seems unrelated to the original issue and needs tests. Can you add this feature in a follow up PR?

File: ibis/expr/types/arrays.py [hunk 4]
```diff
@@ -1131,4 +1131,6 @@ def array(values: Iterable[V]) -> ArrayValue:
     │ [3, 47]              │
     └──────────────────────┘
     """
+    if isinstance(values, ir.ArrayValue):
+        return values
```
Comment (cpcloud) [inline ibis/expr/types/arrays.py:1135]: This functionality seems unrelated to the original issue and needs tests. Can you add this feature in a follow up PR?

File: ibis/expr/types/maps.py [hunk 5]
```diff
@@ -486,6 +484,8 @@ def map(
     │ {'b': 3}             │
     └──────────────────────┘
     """
+    if isinstance(keys, MapValue):
+        return keys
```
Comment (cpcloud) [inline ibis/expr/types/maps.py:488]: This functionality seems unrelated to the original issue and needs tests. Can you add this feature in a follow up PR?

**Review by cpcloud — CHANGES_REQUESTED** [review]

File: ibis/expr/types/structs.py (hunk 3 above)
Comment (NickCrews) [reply ibis/expr/types/structs.py:81]: Oops this was originally needed but then I adjusted the implementation and never reverted this. Will do. I think it's a good change though.

Comment (NickCrews) [reply ibis/expr/types/structs.py]: moved to #8666

File: ibis/backends/duckdb/tests/test_client.py [hunk 6]
```diff
@@ -276,3 +276,12 @@ def test_invalid_connect(tmp_path):
     url = f"duckdb://{tmp_path}?read_only=invalid_value"
     with pytest.raises(ValueError):
         ibis.connect(url)
+
+
+@pytest.mark.xfail(
+    raises=(duckdb.InvalidInputException, duckdb.BinderException),
```
Comment (cpcloud) [inline ibis/backends/duckdb/tests/test_client.py:282]: Why are there two exceptions here?

File: ibis/backends/postgres/compiler.py [hunk 7]
```diff
@@ -330,7 +330,10 @@ def visit_ToJSONArray(self, op, *, arg):
         )
 
     def visit_Map(self, op, *, keys, values):
-        return self.f.map(self.f.array(*keys), self.f.array(*values))
+        # map(["a", "b"], NULL) results in {"a": NULL, "b": NULL} in regular postgres,
```
Comment (cpcloud) [inline ibis/backends/postgres/compiler.py:333]: Wow, that's ... weird.

File: ibis/backends/tests/test_map.py [hunk 8]
```diff
@@ -26,6 +26,29 @@
...
+@pytest.mark.parametrize(
+    "kv",
+    [
+        param((["a", "b"], None), id="null_values"),
+        param((None, ["c", "d"]), id="null_keys"),
+        param((None, None), id="null_both"),
+    ],
+)
+def test_map_nulls(con, kv):
+    k, v = kv
```
Comment (cpcloud) [inline ibis/backends/tests/test_map.py:45]:
```suggestion
    ("k", "v"),
    [
        param(["a", "b"], None, id="null_values"),
        param(None, ["c", "d"], id="null_keys"),
        param(None, None, id="null_both"),
    ],
)
def test_map_nulls(con, k, v):
```

Let pytest do the work for you :)

File: ibis/backends/duckdb/tests/test_client.py [hunk 9]
```diff
@@ -276,3 +276,12 @@ def test_invalid_connect(tmp_path):
...
+    reason="https://github.com/ibis-project/ibis/issues/8632",
+)
+def test_map_null_workaround_xfail():
+    conn = duckdb.connect()
+    conn.sql("SELECT MAP([0, 1, 2], NULL::INT[]);").fetchone()
```
Comment (cpcloud) [inline ibis/backends/duckdb/tests/test_client.py:287]:
```suggestion
def test_map_null_workaround_xfail(con):
    con.con.sql("SELECT MAP([0, 1, 2], NULL::INT[]);").fetchall()
```

File: ibis/backends/duckdb/tests/test_client.py (hunk 6 above)
Comment (NickCrews) [reply ibis/backends/duckdb/tests/test_client.py:282]: Depending on the python/duckdb version you get different errors, was running into both in CI. I wasn't able to actually figure out the exact pattern, I couldnt repro one of the errors locally.  You want a comment for this? Actually figure it out?

File: ibis/backends/postgres/compiler.py (hunk 7 above)
Comment (NickCrews) [reply ibis/backends/postgres/compiler.py:333]: I didn't look more into this, but I'm wondering if this allows you to use Maps as a HashSet, which actually is pretty cool. Maybe we even want to change ibis semantics to allow this?? Idk haven't thought how that would interact with other backends

File: ibis/backends/duckdb/tests/test_client.py (hunk 9 above)
Comment (NickCrews) [reply ibis/backends/duckdb/tests/test_client.py]: thanks, done.

File: ibis/backends/tests/test_map.py (hunk 8 above)
Comment (NickCrews) [reply ibis/backends/tests/test_map.py]: thanks, done.

File: ibis/backends/duckdb/tests/test_client.py (hunk 6 above)
Comment (NickCrews) [reply ibis/backends/duckdb/tests/test_client.py]: OK I pushed with this xfail commented out so you can see [CI failures](https://github.com/ibis-project/ibis/actions/runs/8381568818?pr=8649):

- "duckdb + GeoSpatial *" tests raises BinderErrors
- the other duckdb tests raise InvalidInputExceptions

Comment (NickCrews) [reply ibis/backends/duckdb/tests/test_client.py]: @cpcloud gentle nudge here :)

Comment (cpcloud) [reply ibis/backends/duckdb/tests/test_client.py]: Sorry I missed this!

Totally fine to keep both exception types.

File: ibis/backends/postgres/compiler.py (hunk 7 above)
Comment (cpcloud) [reply ibis/backends/postgres/compiler.py:333]: interesting idea, but let's leave it for later :)

**Review by cpcloud — APPROVED** [review]

**Pull request merged**